### PR TITLE
Gnuplot (GASTON) backend testing

### PR DESCRIPTION
**Pull request by isentropic**
Dear all, 
here is the first attempt at making gnuplot (gaston) plots backend. Still very early stage, still testing, too many things remain, but it feels like gnuplot can be a good backend as there have been no hacks implementing it so far.

Please test it like  (naturally need to have gaston and gnuplot working beforehand)

```
using Plots; gaston()
plot(1:10, xlab="gnuplot is not bad at all", xguidefontsize=15, ylab="notbad", yguidefontsize=5, tick_direction=:out, guidefontfamily="Serif")
```
image

```
plot(
       plot(1:10, xlab="gnuplot is not bad at all", xguidefontsize=15, ylab="notbad", yguidefontsize=5, tick_direction=:out, guidefontfamily="Serif"), plot(sin.(1:100))
       )
```
image

Not many things work, in fact, most don't. But I think there is potential to this as it behaved very nice so far. Only tested in repl, but jupyter has to work too. Waiting for reviews from @mbaz 
q
```

### CODE REVIEW

File: src/backends/gaston.jl [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+# Create the window/figure for this backend.
+function _create_backend_figure(plt::Plot{GastonBackend})
+    state_handle = G.nexthandle() # for now all the figures will be kept
+    plt.o = G.newfigure(state_handle)
+end
+
+
+function _before_layout_calcs(plt::Plot{GastonBackend})
+    # Initialize all the subplots first
+    plt.o.subplots = G.SubPlot[]
```
Comment (isentropic) [inline src/backends/gaston.jl:16]: clear the default subplot

Comment (isentropic) [reply src/backends/gaston.jl:16]: all subplots are added next

File: src/backends/gaston.jl [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    grid = size(plt.layout)
+    plt.o.layout = grid
```
Comment (isentropic) [inline src/backends/gaston.jl:19]: I think gnuplot can only support simple (column, row) grid

File: src/backends/gaston.jl [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+end
+
+function _update_min_padding!(sp::Subplot{GastonBackend})
+    sp.minpad = (20mm, 5mm, 2mm, 10mm)
+end
+
+function _update_plot_object(plt::Plot{GastonBackend})
+end
+
+function _show(io::IO, ::MIME"image/png", plt::Plot{GastonBackend})
```
Comment (isentropic) [inline src/backends/gaston.jl:38]: saving figures would be done through this function

File: src/backends/gaston.jl [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+end
+
+function gaston_add_series(plt::Plot{GastonBackend}, series::Series)
+    # Gaston.Curve = Plots.Series
+    sp = series[:subplot]
+    g_sp = sp.o  # Gaston subplot object
+
+    seriesconf = gaston_parse_series_args(series)  # Gnuplot string
+    c = G.Curve(series[:x], series[:y], nothing, nothing, seriesconf )
```
Comment (isentropic) [inline src/backends/gaston.jl:64]: G.Curve is almost the same as plots series

File: src/backends/gaston.jl [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    end
+
+    # Then add the series (curves in gaston)
+    for series in plt.series_list
+        gaston_add_series(plt, series)
+    end
+end
+
+function _update_min_padding!(sp::Subplot{GastonBackend})
+    sp.minpad = (20mm, 5mm, 2mm, 10mm)
```
Comment (BeastyBlacksmith) [inline src/backends/gaston.jl:32]: I find it highly confusing if a backend sets these minpads

Comment (isentropic) [reply src/backends/gaston.jl:32]: This callback is meant for the backend to adjust the padding. If you cannot adjust all the padding right away, because Plots calculates the padding when grid is complex.

With a complex layout, backend cannot align all the subplots itself, Plots has the aligning logic, so it this callback your are supposed to add the padding Plots tells you to. 

For the moment I do not know if it is possible to communicate with gnuplot to estimate the paddings

Comment (isentropic) [reply src/backends/gaston.jl:32]:
```
l = @layout([a{0.1h}; b [c; d e]])
plot(randn(100, 5), layout = l, t = [:line :histogram :scatter :steppre :bar], leg = false, ticks = nothing)
```
image

Look how pyplot aligns http://docs.juliaplots.org/latest/generated/pyplot/#pyplot-ref16 this so well, the axes spines are well aligned even if you put the uneven sized labels (guides).

 It is only possible with this callback, Plots basically calculates needed padding for each subplot in layouts.jl, it is a fairly complex recursive algorithm. First you put all the subplots, tell how much *at least* should each subplot have padding. Then plot tells you how much exactly each subplot would need for alignment.
So basically you should supply the minimal padding, then Plots in this callback tells you how much it should be (more than minimal to align with other subplots)

**Comment by isentropic** [comment]
More things work now
```
linetypes = [:path :steppre :steppost :sticks :scatter]
n = length(linetypes)
x = Vector[sort(rand(20)) for i = 1:n]
y = rand(20, n)
plot(x, y, line = (linetypes, 3), lab = map(string, linetypes), ms = 15)
```
image

**Comment by isentropic** [comment]
@mbaz  I wanted to know if communication with gnuplot is possible? Basically
`show` commands in gnuplot repl

File: src/backends/gaston.jl (hunk 5 above)
Comment (BeastyBlacksmith) [reply src/backends/gaston.jl:37]: But isn't it also causing unwanted whitespace, when you e.g. set `margin = 0Plots.mm` ?

Comment (isentropic) [reply src/backends/gaston.jl:37]: The key word margins are extra margins from users, ideally there should be no need for it at all. But with colorbars and such setting extra margin is a hack, it should've been accounted for by plots

**Comment by mbaz** [comment]
@isentropic If you mean running arbitrary gnuplot commands, it's planned for the next version (https://github.com/mbaz/Gaston.jl/issues/154)

As for `show` commands: Gaston (in my private devel branch, which will become 2.0) supports showing ascii, svg, png, sixelgd and canvas on the terminal (as text). Is that what you mean?

**Comment by isentropic** [comment]
http://www.gnuplot.info/docs_5.4/Gnuplot_5_4.pdf 
show command as in the docs. So back communication from gnuplot basically

**Comment by mbaz** [comment]
Oh! I thought you meant Julia's `show`, not gnuplot's. Sorry for the confusion.

This is not supported at the moment -- you can send commands to gnuplot using `Gaston.gnuplot_send()`, but Gaston does not read back the result. I plan to support this in the next version.

**Comment by isentropic** [comment]
New stuff like axis keywords and stuff added. Comments are welcome
```
plot(histogram(randn(10000), yscale=:log10), plot(1:500))
```
image

**Comment by isentropic** [comment]
Supported args and markers are written in `backends.jl` I think that this might be ready for some tests now. Can someone help do it? I was never good with tests and how to run them. 
Also, @mbaz would you please write up the `save` function? I'm not sure what would be the right way to export the figures. I think it would be fairly easy, similar to `display` function

File: src/backends/gaston.jl (hunk 3 above)
Comment (isentropic) [reply src/backends/gaston.jl:43]: @mbaz it would be really helpful if you could help me fill this cuntion correctly

**Comment by isentropic** [comment]
how do i make this a draft? I presed complete review now its green again

**Comment by daschw** [comment]
I tested it and it works amazing (from the REPL) so far. Really nice effort! What are your plans? Do you want to keep working on this PR to add more features, or would you rather merge a first basic PR and continue working on new features in other PRs?

**Comment by mbaz** [comment]
@isentropic What do you mean about the `save` function? It's already there: https://github.com/mbaz/Gaston.jl/blob/master/src/gaston_save.jl https://mbaz.github.io/Gaston.jl/stable/introduction/#Saving-plots-1

Awesome work BTW!

**Comment by BeastyBlacksmith** [comment]
> how do i make this a draft? I presed complete review now its green again

"Convert to draft" is below the reviewer section

File: src/backends/gaston.jl [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        )
+        while !isfile(tmpfile) end  # avoid race condition with read in next line
+        write(io, read(tmpfile))
+        rm(tmpfile, force=true)
+        nothing
+
+    end
+end
+
+function _show(io::IO, mime::MIME{Symbol("image/png")},
```
Comment (isentropic) [inline src/backends/gaston.jl:78]: @mbaz Image/png I believe is for png figures, as gnuplot has a constant dpi I had add some scaling hacks here

File: src/backends/gaston.jl [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    "application/eps"         => "epscairo",   # NEED fixing TODO
+    "image/eps"               => "epslatex",   # NEED fixing TODO
+    "application/pdf"         => "pdfcairo",   # NEED fixing TODO
+    "application/postscript"  => "postscript", # NEED fixing TODO
+    "image/svg+xml"           => "svg",
+    "text/latex"              => "tikz",       # NEED fixing TODO
+    "application/x-tex"       => "epslatex",   # NEED fixing TODO
+    "text/plain"              => "dumb",       # NEED fixing TODO
+)
+    @eval function _show(io::IO, ::MIME{Symbol($mime)}, plt::Plot{GastonBackend})
```
Comment (isentropic) [inline src/backends/gaston.jl:57]: @mbaz, most of the terminals do not work properly and I do not know why. PDF for some reason produces weird output. Also, latex terminals complain about the file format.

**Review by isentropic — COMMENTED** [review]
@mbaz Would you please look at these comments? For some reason the saving for pdf, latex, and txt works unexpectedly

**Comment by isentropic** [comment]
@daschw Yeah, I think I need to polish up saving and some more plotting args like plot title and legend box positioning. Then perhaps it would cover the bare minimum for simple 2d plots. Later, I'm planning to add colorbar and 3d plots.

**Comment by isentropic** [comment]
`p = plot(rand(50,5), fontfamily="Serif", legendtitle="legend", leg=:outerbottomleft, title="titles are here")`
image

a few more things work, I guess right now it is minimal working verion

**Comment by clad26** [comment]
@isentropic Great work!
Any idea how long would it take to get it as functional as other backends?
Like, what percent do you think can Gaston backend provide for now, compared to the complete product?
Again, thanks for taking the time!

**Comment by isentropic** [comment]
@clad26 I really wanted to have gaston as an invitation for new contributors really. It is an inferior backend that is very lean in codebase (300 lines only). I think people familiar with gnuplot in general will find it easy to contribute and improve on this. Right I would say that a lot of 2d plots work already, but 3d support is not here yet.

**Comment by clad26** [comment]
> @clad26 I really wanted to have gaston as an invitation for new contributors really. It is an inferior backend that is very lean in codebase (300 lines only). I think people familiar with gnuplot in general will find it easy to contribute and improve on this. Right I would say that a lot of 2d plots work already, but 3d support is not here yet.

Does it provide histogram yet?

**Comment by isentropic** [comment]
Yeah it works

**Comment by isentropic** [comment]
@clad26 if you are interested in this, let me know what features do you need the most

**Comment by VarLad** [comment]
@isentropic It'd be nice if this can be fully completed and merged as a part of plots
It might be the fastest plotting backend of all 
Since the artifacts are almost here, this might be good

You can tell me from where can I start implementing 3D and what in 2D is left, so that I can help too?

**Comment by isentropic** [comment]
The important think that prevents me from merging this, is the lack of testing and community feedback. This was an experiment of mine to see if gnuplot can even work as a backend for plots. I'd hold off with implementing 3d. We need to test 2d suite for egregious bugs first (I cannot afford to do testing --- that is to use this on daily basis and see what is missing). 

Regarding comPat, I'm not sure how to set this up, perhaps you could send a PR to this fork?

If you could assist with this backend, I'd be happy to see contribution

**Comment by t-bltg** [comment]
@isentropic, we now have Gaston in the docs: https://docs.juliaplots.org/latest/generated/gaston :wink:
